src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITValue.cpp

```diff
@@ -262,6 +471,65 @@ JITValuePointer LLVMJITValue::add(JITValue& rh) {
       getValueSubTypeTag());
 }
 
+JITValuePointer LLVMJITValue::add_with_error_check(JITValue& rh) {
+  // only integer type need overflow check
+  auto lhs_llvm_type = getLLVMType(getValueTypeTag(), parent_function_.getLLVMContext());
+  if (!lhs_llvm_type->isIntegerTy()) {
+    return add(rh);
+  }
+
+  LLVMJITValue& llvm_rh = static_cast<LLVMJITValue&>(rh);
+  if (JITTypeTag::POINTER == getValueTypeTag()) {
+    auto rh_type = rh.getValueTypeTag();
+    if (rh_type != JITTypeTag::INT8 && rh_type != JITTypeTag::INT16 &&
+        rh_type != JITTypeTag::INT32 && rh_type != JITTypeTag::INT64) {
+      LOG(FATAL) << "Invalid index type for pointer offset operation. Name="
+                 << rh.getValueName() << ", Type=" << getJITTypeName(rh_type);
+    }
+  } else {
+    checkOprandsType(this->getValueTypeTag(), rh.getValueTypeTag(), "add");
+  }
+
+  llvm::BasicBlock* check_ok{nullptr};
+  llvm::BasicBlock* check_fail{nullptr};
+
+  check_ok = llvm::BasicBlock::Create(
+      parent_function_.getLLVMContext(), "ovf_ok", &parent_function_.func_);
+  check_fail = llvm::BasicBlock::Create(
+      parent_function_.getLLVMContext(), "ovf_detected", &parent_function_.func_);
+
+  // get compute function
+  auto llvm_module = parent_function_.func_.getParent();
+  auto func = llvm::Intrinsic::getDeclaration(
+      llvm_module, llvm::Intrinsic::sadd_with_overflow, lhs_llvm_type);
+
+  // compute result and overflow
+  auto ret_and_overflow =
+      getFunctionBuilder(parent_function_)
+          .CreateCall(func, std::vector<llvm::Value*>{load(), llvm_rh.load()});
+  auto ret = getFunctionBuilder(parent_function_)
+                 .CreateExtractValue(ret_and_overflow, std::vector<unsigned>{0});
+  auto overflow = getFunctionBuilder(parent_function_)
+                      .CreateExtractValue(ret_and_overflow, std::vector<unsigned>{1});
+
+  // return error
+  getFunctionBuilder(parent_function_).CreateCondBr(overflow, check_fail, check_ok);
+  getFunctionBuilder(parent_function_).SetInsertPoint(check_fail);
+  getFunctionBuilder(parent_function_)
+      .CreateRet(llvm::ConstantInt::get(
+          llvm::Type::getInt32Ty(parent_function_.getLLVMContext()),
+          ERROR_CODE::ERR_OVERFLOW_OR_UNDERFLOW));
+  getFunctionBuilder(parent_function_).SetInsertPoint(check_ok);
```

Comment:
Seems all error check procedures are based on IF control flow, I think ```IfBuilder``` in JITLib could be applied to simplify the code and keep consistency. 

BTW, for ```mod_with_error_check``` and ```div_with_error_check```, there are lots of code duplication with ```mod``` and ```div```, any chance to eliminate them?